tools/clang/test/HLSLFileCheck/validation/CallableParamIsStruct.ll

```diff
@@ -0,0 +1,33 @@
+; RUN: %dxv %s | FileCheck %s
+
+; CHECK: Function: BadCallable: error: Argument 'f' must be a struct type for callable shader function 'BadCallable'.
+
+; Test based on IR generated from the following HLSL:
+; struct Param { float f; };
+; [shader("callable")]
+; void CallableProto(inout Param p) { p.f += 1.0; }
+;
+; export void BadCallable(inout float f) { f += 1.0; }
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:32-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64"
+target triple = "dxil-ms-dx"
+
+; Function Attrs: nounwind
+define void @BadCallable(float* noalias nocapture dereferenceable(4) %f) #0 {
+  ret void
+}
+
+attributes #0 = { nounwind }
+
+!dx.version = !{!0}
+!dx.valver = !{!1}
+!dx.shaderModel = !{!2}
+!dx.entryPoints = !{!3, !4}
+
+!0 = !{i32 1, i32 3}
+!1 = !{i32 1, i32 7}
```

Comment:
This is the validator version.  This value (1.7) a problem, since it will restrict the test to working when run against validator version 1.7 (the most recently released version).

With tests in `ValidationTest.cpp` that use the compile->disassemble->modify->assemble->validate path, the validator version would be set to the currently attached one (including the version from prior releases of external validators, against which we run tests regularly).  This would appropriately impact IR (metadata) in certain key cases.

With tests written in IR, we would likely need to permute this with different validator versions for some tests.  For most tests, we can probably get away with setting this to the minimum validator version that supported the rule we are testing.  For instance, I think that would be 1.3 for this test instead of 1.7.

However, just setting this to a minimum version may not be sufficient for some tests, since there can be different decisions/allowances in the validator depending on whether the IR was originally targeting an older validator version or the current one being run.  We may also need to test different DXIL versions for some scenarios.

Permuting versions within a single test can't quite be done unless we have a way to replace certain IR input elements with variables from a RUN line.  If that was supported, we could use different RUN lines, different `%dxilver` requirements for each, and different `-D` variables for FileCheck.